### ISSUE TITLE
Windows: Stack now considers GHC 8.6.4 invalid as no haddock.exe

### ISSUE THREAD

**Issue by mpilgrem** [commercialhaskell/stack#4727]
On Windows 10, `stack` version `Version 1.10.0, Git revision abf4ca08706e02256a500e507ff7ec47c21e55c8 (7365 commits) x86_64 hpack-0.31.1` now considers GHC 8.6.4 (provided with stack) to be invalid because the GHC binaries do not come with `haddock.exe` (they come only with `haddock-8.6.4.exe`).

See the contents of `ghc-8.6.4-x86_64-unknown-mingw32.tar.xz\ghc-8.6.4-x86_64-unknown-mingw32.tar\ghc-8.6.4\bin\` for the missing `haddock.exe` (https://www.haskell.org/ghc/download_ghc_8_6_4.html#windows64)

For a plain vanilla project (`stack new test`, `cd test`):

~~~~
>stack build
Found an invalid compiler at "C:\\Users\\mikep\\AppData\\Local\\Programs\\stack\\x86_64-windows\\ghc-8.6.4\\bin\\ghc.EXE": Control.Exception.Safe.throwString called with:

Could not find any of: ["C:\\Users\\mikep\\AppData\\Local\\Programs\\stack\\x86_64-windows\\ghc-8.6.4\\bin\\haddock.exe","C:\\Users\\mikep\\AppData\\Local\\Programs\\stack\\x86_64-windows\\ghc-8.6.4\\bin\\haddock-ghc.exe"]
Called from:
  throwString (src/Stack\Setup.hs:733:25 in stack-1.10.0-EqoqGlQCHrZF4lstia6Gor:Stack.Setup)
~~~~

This may be an upstream issue if GHC 8.6.4 was meant to ship with `haddock.exe` as well as `haddock-8.6.4.exe`.

**Comment by snoyberg** [comment]
Thanks for reporting! Would you be able to test out commit 59c415171c8ee3c6872ce647955bd0fabb241500 of Stack and see if that resolves the issue?

**Comment by mpilgrem** [comment]
Stack (version: `Version 1.10.0, Git revision 59c415171c8ee3c6872ce647955bd0fabb241500 (7366 commits) x86_64 hpack-0.31.1`) does eliminate the error on `stack build`.

I have `haddock.exe` elsewhere on the path, but `stack build --haddock --haddock-arguments="-odocs"` ultimately fails (after creating documentation using that 'other' `haddock.exe`) with (extracts only):
~~~~
>stack build --haddock --haddock-arguments="-odocs"
...
Preprocessing library for test-0.1.0.0..
Warning: The documentation for the following packages are not installed. No
Running Haddock on library for test-0.1.0.0..
links will be generated to these packages: base-4.12.0.0, ghc-prim-0.5.3,
integer-gmp-1.0.2.0
Haddock coverage:
   0% (  0 /  2) in 'Lib'
  Missing documentation for:
    Module header
    someFunc (src\Lib.hs:5)
Warning: Lib: could not find link destinations for:
    IO
Documentation created: .stack-work\dist\e626a42b\doc\html\test\index.html,
.stack-work\dist\e626a42b\doc\html\test\test.txt
...
Registering library for test-0.1.0.0..
Updating Haddock index for local packages in
C:\Users\mikep\Documents\Code\Haskell\test\.stack-work\install\a56124d7\doc\index.html
Received ExitFailure 1 when running
Raw command: "C:\\Users\\mikep\\AppData\\Local\\Programs\\stack\\x86_64-windows\\ghc-8.6.4\\bin\\haddock-8.6.4.exe" "--optghc=-package-db=C:\\sr\\snapshots\\ebc9c4d7\\pkgdb" "--optghc=-package-db=C:\\Users\\mikep\\Documents\\Code\\Haskell\\test\\.stack-work\\install\\a56124d7\\pkgdb" -odocs --gen-contents --gen-index -i "C:\\Users\\mikep\\Documents\\Code\\Haskell\\test\\.stack-work\\install\\a56124d7\\doc\\test-0.1.0.0,C:\\Users\\mikep\\Documents\\Code\\Haskell\\test\\.stack-work\\install\\a56124d7\\doc\\test-0.1.0.0\\test.haddock"
Run from: C:\Users\mikep\Documents\Code\Haskell\test\.stack-work\install\a56124d7\doc\
Standard error:

Unable to start C:\Users\mikep\AppData\Local\Programs\stack\x86_64-windows\ghc-8.6.4\bin/haddock.exe (error code: 2)
~~~~

I also investigated further. Looking at GHC 8.6.3 and earlier by analogy, given very different file sizes, `haddock-8.6.4.exe` (small file) must do something quite different from `haddock.exe` (large file). I suspect the 'missing' `haddock.exe` from GHC 8.6.4's `\bin` folder is simply an upstream mistake. My 'fix' was to copy GHC 8.6.3's `haddock.exe` into GHC 8.6.4's `\bin` folder.

**Comment by snoyberg** [comment]
There are some inherent problems with the GHC 8.6.4 Windows bindist, especially around `haddock`. We're not going to be able to fix that at the Stack level, and instead we'll need to wait for GHC 8.6.5 to be released. My goal here is to fix Stack master so that it's at least able to build code with GHC 8.6.4 on Windows.
